routing/cross_mwm_index_graph.cpp

```diff
@@ -41,45 +80,106 @@ bool CrossMwmIndexGraph::IsTransition(Segment const & s, bool isOutgoing)
+
+  auto it = m_transitionCache.find(s.GetMwmId());
+  CHECK(it != m_transitionCache.cend(), ());
+
+  ms::LatLon const & latLog = isOutgoing ? GetLatLon(it->second.m_outgoing, s)
+                                         : GetLatLon(it->second.m_ingoing, s);
+  for (string const & name : neighboringMwm)
+  {
+    NumMwmId const numMwmId = m_numMwmIds->GetId(CountryFile(name));
+    TRoutingMappingPtr mappingPtr = GetRoutingMapping(numMwmId);
+    if (mappingPtr == nullptr)
+      continue; // mwm was not loaded.
+
+    if (isOutgoing)
+    {
+      mappingPtr->m_crossContext.ForEachIngoingNodeNearPoint(latLog, [&](IngoingCrossNode const & node){
+        AddTransitionSegment(mappingPtr->m_segMapping, node.m_nodeId, numMwmId, twins);
+      });
+    }
+    else
+    {
+      mappingPtr->m_crossContext.ForEachOutgoingNodeNearPoint(latLog, [&](OutgoingCrossNode const & node){
+        AddTransitionSegment(mappingPtr->m_segMapping, node.m_nodeId, numMwmId, twins);
+      });
+    }
+  }
+
+  my::SortUnique(twins);
 }
 
 void CrossMwmIndexGraph::GetEdgeList(Segment const & /* s */,
-                                         bool /* isOutgoing */, std::vector<SegmentEdge> & /* edges */) const
+                                     bool /* isOutgoing */, std::vector<SegmentEdge> & /* edges */) const
 {
   // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
   // and if so to use it. If not, osrm cross mwm sections should be used.
   NOTIMPLEMENTED();
 }
+
+TRoutingMappingPtr CrossMwmIndexGraph::GetRoutingMapping(NumMwmId numMwmId)
+{
+  CountryFile const & countryFile = m_numMwmIds->GetFile(numMwmId);
+  TRoutingMappingPtr mappingPtr = m_indexManager.GetMappingByName(countryFile.GetName());
+  CHECK(mappingPtr, ("countryFile:", countryFile));
+  if (!mappingPtr->IsValid())
+    return nullptr; // mwm was not loaded.
+  mappingPtr->LoadCrossContext();
+  return mappingPtr;
+}
+
+void CrossMwmIndexGraph::Clear()
+{
+  m_transitionCache.clear();
+}
+
+void CrossMwmIndexGraph::FillWholeMwmTransitionSegments(NumMwmId numMwmId)
+{
+  if (m_transitionCache.count(numMwmId) != 0)
+    return;
+
+  TRoutingMappingPtr mappingPtr = GetRoutingMapping(numMwmId);
+  if (mappingPtr == nullptr)
+    return;
+
+  TransitionSegments transitionSegments;
+  mappingPtr->m_crossContext.ForEachOutgoingNode([&](OutgoingCrossNode const & node)
+  {
+    FillTransitionSegments(mappingPtr->m_segMapping, node.m_nodeId, numMwmId,
+                           node.m_point, transitionSegments.m_outgoing);
+  });
+  mappingPtr->m_crossContext.ForEachIngoingNode([&](IngoingCrossNode const & node)
+  {
+    FillTransitionSegments(mappingPtr->m_segMapping, node.m_nodeId, numMwmId,
+                           node.m_point, transitionSegments.m_ingoing);
+  });
+  auto const p = m_transitionCache.emplace(numMwmId, move(transitionSegments));
+  CHECK(p.second, ("Mwm num id:", numMwmId, "has been inserted before. Country file name:",
```

Comment:
No need in CHECK - it's ensured in the beginning of the function. ASSERTION is enough.